32 introduce 3mf based geometry build based on db

jhajduga opened the pull request.

jhajduga committed:
Positiopning from CAD model.
Almost isocentre where it should be...
Almost.

jhajduga committed:
Positioning, renaming and if iba then X and Z in cell positioning are multiplied by -1

barachwal committed:
put iba rotaion check directly in layer

barachwal committed:
remove redundant code and tune naming

jhajduga committed:
WIP -> Iba update and naming in code.

TODO!
Change All Plans (Positioning from CSV move to to D3DDetector_Detector)
TopPositionInenv rename to TranslationInLocalFrame
TranslationInLocalFrame moved to Cell
Uptade fobs to incorporate that.

jhajduga committed:
WIP

jhajduga committed:
There were a doubled shift of the phantom. Partialy resolved.

TODO - update cell positioning to work properly...

barachwal committed:
update to current develop

barachwal committed:
iba positioning - done!

jhajduga committed:
Updated - Positioning for IbaImRT Phantom

jhajduga committed:
Proper box positioning and plan.
Proper table rotation.
Proper box rotation.

jhajduga committed:
Static radiotherapy-ready setup for Iba ImRT Just Box.

jhajduga committed:
Updated - enable proper simplified MLC positioning.

barachwal committed:
Merge branch 'develop' into 42-get-cell-positioning-from-new-sc-mapping-db

barachwal committed:
renamings

jhajduga committed:
Proper MLC positioning and orientation.

jhajduga committed:
Merge remote-tracking branch 'origin/42-get-cell-positioning-from-new-sc-mapping-db' into 42-get-cell-positioning-from-new-sc-mapping-db

barachwal committed:
fix conficts

barachwal committed:
fix d3dtray issue

jhajduga committed:
WIP.

TODO:
Incorporate it to word construction.

barachwal committed:
works for StlDetectorWithPositioningFromCsv

barachwal committed:
works for PositioningFromCsv

barachwal committed:
WIP ComputeCellsPositioning

jhajduga committed:
Modular Detector should inherit from IPhysicalVolume and not VPatent
Todo - fix it

jhajduga committed:
Uptade - Geo creation from 3mf format (saved to db with vertexes in csv) is working almost properly.

Todo:
Deal with the problem in which some walls are “too small” to be considered valid.

fCarTolerance = myMaxExtent * 1e-11 * mm
This depends on the size of the world.
The larger the world, the larger you must have geometric elements in the simulation.

jhajduga committed:
Phantom Explosion state fixed.

barachwal committed:
make it working for regular positioning

barachwal committed:
remove Layer Pads export

barachwal committed:
cleanup

barachwal committed:
rm layer context from d3ddetector

barachwal committed:
rmD3DMLayer from project

barachwal committed:
Merge branch '32-introduce-3mf-based-geometry-build-based-on-db' of github.com:dose3d/g4rt into 32-introduce-3mf-based-geometry-build-based-on-db

jhajduga committed:
WIP - for propouse of merging changes.

jhajduga committed:
Merge branch '32-introduce-3mf-based-geometry-build-based-on-db' of github.com:dose3d/g4rt into 32-introduce-3mf-based-geometry-build-based-on-db

jhajduga committed:
Added tool to change Surface tolerance size.

TODO:

Check in ModularPhantom.cc ModularPhantom::Construct if the triangles are to small or degenerate - if they are - do not create them.

TODO: V2:  During exporting meshes to DB -> simplifie meshes in trimesh.

jhajduga committed:
Code that works well enough.  I am careful not to overcomplicate things.

Temporaly - fewer errors
when prouking geometry.
During simulation, more chance that a particle will get stuck because the few remaining damaged elements are more damaged.

jhajduga committed:
WIP - refac.

jhajduga committed:
temp data added

jhajduga committed:
GeometryBuilder and GeometryParser have been refactoured.
The directory structure has been revised.

TODO:
A lot of

jhajduga committed:
Wip - almost there...

jhajduga committed:
// TODO: Filtrowanie elementów, które chcemy usunąć z IbaImRT
    // TODO: Musi zostać stworzony jakiś modół który pozwoli pobrać ścieżkę do DB i do CSV z z PhantomWorld

Bouth - project them as generic mechanizm and then - implement them in IBA and ModularWP

jhajduga committed:
TODO:

- Rename Geometry Parser to GeometryDBReader.
- Inside GeometryDBReader -> build list of scintilator cells.
- if path to DB exist -> then then do stuff.
- read materials from DB && translate them to geant materials.
- move and rotate all nev elements inside patient world.
- set minimal patient world size based on GetExtent():
    // Get extents
    G4VisExtent ve = solid->GetExtent();
    G4double xmin = ve.GetXmin(), xmax = ve.GetXmax();
    G4double ymin = ve.GetYmin(), ymax = ve.GetYmax();
    G4double zmin = ve.GetZmin(), zmax = ve.GetZmax();

jhajduga committed:
WIP

jhajduga committed:
Cleaning - Bug:

label = D3DDetector_Cell_ON245  position: (58.7,47.95,59.45)
[INFO]:: IPhysicalVolume::Construct: No direct pv to link...
[INFO]:: Link logical D3DDetector_Cell_ON245 with G4PV: phmWorldPV
transformPosition: (58.7,47.95,59.45)...
 no physical volume found.
 no rotation.
 no rotation.

Looking for source of bug and wgy it was working elier...

jhajduga committed:
Almost done...

jhajduga committed:
Working code.
Python intrerpreter finalized properly.

jhajduga committed:
TODO def.

jhajduga committed:
Unnecessary double finalization of GeometryDBReader

jhajduga committed:
Working code.

jhajduga committed:
Add d3df-patients as submodule under submodules/g4rtd3df-patients

jhajduga committed:
Patient update

jhajduga committed:
Capping updated

jhajduga committed:
Submodule update.

jhajduga committed:
Submodule update

jhajduga committed:
Modular Water Phantom updated.

jakuberich commented on core/geometry/Patient/DBGeoBuilder/GeometryBuilder.cc:
```diff
@@ -0,0 +1,85 @@
+
+////////////////////////////////////////////////////////////////////////////////
+///
+GeometryBuilder::GeometryBuilder():TomlConfigModule("GeometryBuilder"){}
```
```suggestion
GeometryBuilder::GeometryBuilder():TomlConfigModule("GeometryBuilder"){

  ParseTomlConfig();
}


////////////////////////////////////////////////////////////////////////////////
///   
void GeometryBuilder::ParseTomlConfig(){
  SetTomlConfigFile();
  auto configFile = GetTomlConfigFile();
  auto configPrefix = GetTomlConfigPrefix();

  // LOGSVC_INFO("Importing configuration from:\n{}",configFile); // Not Logable RN
  // if (!svc::checkIfFileExist(configFile)) {
  //   LOGSVC_CRITICAL("File {} not fount.", configFile);
  //   G4Exception("GeometryBuilder", "ParseTomlConfig", FatalErrorInArgument, "");
  // }

  std::cout << "Importing configuration from:\n" << configFile << "\n";
  auto config = toml::parse_file(configFile);
  std::cout << config << "\n";


  m_centrePositionX = config["Phantom"]["Position"][0].value_or(0.0);
  m_centrePositionY = config["Phantom"]["Position"][1].value_or(0.0);
  m_centrePositionZ = config["Phantom"]["Position"][2].value_or(0.0);

  m_phantomRotationX = config["Phantom"]["Rotation"][0].value_or(0.0);
  m_phantomRotationY = config["Phantom"]["Rotation"][1].value_or(0.0);
  m_phantomRotationZ = config["Phantom"]["Rotation"][2].value_or(0.0);

  m_exclusde_object_list = config["ExcludeObjList"].as_array();

  if (m_exclusde_object_list) {
    for (const auto& elem : *m_exclusde_object_list) {
      if (elem.is_string()) {
        std::string value = elem.value_or("");
        std::cout << "Wykluczony obiekt: " << value << "\n";
      }
    }
  }
}
```

jakuberich commented on core/geometry/Patient/DBGeoBuilder/GeometryBuilder.cc:
```diff
@@ -0,0 +1,85 @@
+  
+  
+  ////////////////////////////////////////////////////////////////////////////////
+  /// 
```
```suggestion
    
  
```

jhajduga committed:
WIP

jhajduga committed:
Excluding components from Toml -> Done.

jhajduga committed:
TODO:

 - Update README.md here and in the submodule.
 - Implement rotation of 3mf parts.

jhajduga committed:
README Uptade

jhajduga committed:
Submodule update

jhajduga committed:
Submodule uptaded

jhajduga committed:
Jobs for testbeam

jhajduga committed:
HDF5 old dependencies remopved, Annoying "make" warning about visability removed, UPdated - g4rt_env.yml to working wersion (also with tools to visualize phantom by trimesh).

jhajduga committed:
Pyglet and fretype update.

jhajduga committed:
Properly rotated 3mf geometry.

jhajduga committed:
PDD update

barachwal committed:
update patient reference

barachwal committed:
add some improvements

barachwal committed:
fix and patient update

barachwal committed:
update patient sumbmodule

jhajduga committed:
Cmake uptades - python path issue.

jhajduga committed:
Last needed changes.
(missing collon, GDML language (on polisk systems))

barachwal committed:
55 refactor volume calc for the analysis (#60)

* add volume for vpatient

* add extra cell volume logic

* fix CellInfo ctr issue

* remove ::SIZE usage

* refactor ::SIZE type

barachwal committed:
rm obsolete scoring func

barachwal committed:
touch gtest

barachwal commented:
Podczas importu geometrii mogą występować liczby zmiennoprzecinkowe bardzo bliskie zeru, które wynikają z niedokładności obliczeń numerycznych (np. `-1.02318e-12`). W celu zapewnienia spójności danych geometrii i uniknięcia błędów obliczeniowych, należy zaokrąglać wszystkie takie wartości do zera, jeśli ich bezwzględna wartość jest mniejsza od zdefiniowanego globalnie progu dokładności `GEOM_EPSILON`.

**Wymagania:**

* Zdefiniuj globalną stałą `GEOM_EPSILON` (np. `1e-6`).
* Zaimplementuj funkcję `sanitizeZero(double value)`, która:

  * przyjmuje wartość zmiennoprzecinkową,
  * zwraca `0.0`, jeśli `|value| < GEOM_EPSILON`,
  * w przeciwnym razie zwraca wartość bez zmian.

**Przykład:**

```cpp
constexpr double GEOM_EPSILON = 1e-6;

double svc::sanitizeZero(double value, double treshold=GEOM_EPSILON) {
    return std::abs(value) < GEOM_EPSILON ? 0.0 : value;
}
```

**Zastosowanie:**

Używaj funkcji `sanitizeZero()` we wszystkich miejscach importu i przetwarzania danych geometrii, gdzie dokładność numeryczna może wpływać na wynik.

barachwal committed:
add cad export issue fix

barachwal committed:
GeometryBuilder cleanup
